### PR TITLE
Skip update service, if no difference in specified changes.

### LINKED ISSUE #25770: service update "--mount-add" not idempotent
Using `docker service update --mount-add` updates the service even if there's no difference in the specified option. (This may be related to other issues where defaults are not compared correctly?)

To reproduce;

``` bash
$ docker service create --name foo --mount type=volume,destination=/foo nginx:alpine
```

Verify a single container (and volume) is created:

```
$ docker service ps foo

ID                         NAME   IMAGE         NODE          DESIRED STATE  CURRENT STATE           ERROR
92t1bjd925311lxslbi0bxwrv  foo.1  nginx:alpine  7e44720a5b2a  Running        Running 46 seconds ago

$ docker volume ls

DRIVER              NAME
local               ab37ce0a9bcf14355e10292638ad9c38cdc04306d056533302f3b13c48ebe15b
```

Then update the service, using the exact same options:

``` bash
$ docker service update --mount-add type=volume,destination=/foo foo
foo
```

As can be seen, a new task is created, and a new volume:

``` bash
$ docker service ps foo

ID                         NAME       IMAGE         NODE          DESIRED STATE  CURRENT STATE            ERROR
cbz5394monlimlnm0ruhq9m1u  foo.1      nginx:alpine  7e44720a5b2a  Running        Running 41 seconds ago
92t1bjd925311lxslbi0bxwrv   \_ foo.1  nginx:alpine  7e44720a5b2a  Shutdown       Shutdown 44 seconds ago

$ docker volume ls

DRIVER              NAME
local               2628abbf4b11b7176a044da2e251dc3bb9c87a5e9a609024c0752577f9e2a110
local               ab37ce0a9bcf14355e10292638ad9c38cdc04306d056533302f3b13c48ebe15b
```

Tested on current master (with some changes to 'docs', hence `-unsupported`;

```
Client:
 Version:      1.13.0-dev
 API version:  1.25
 Go version:   go1.6.3
 Git commit:   19f8c58-unsupported
 Built:        Tue Aug 16 15:37:58 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0-dev
 API version:  1.25
 Go version:   go1.6.3
 Git commit:   19f8c58-unsupported
 Built:        Tue Aug 16 15:37:58 2016
 OS/Arch:      linux/amd64
```

### PR DESCRIPTION
**Pull request by kunalkushwaha**
**\- What I did**
Check for changes in current and new spec, which is build in result of `docker service update`.
If there is no changes, Skip the update.

NOTE: This fix is depended on PR #25974

**\- How I did it**
in function [UpdateService](daemon/cluster/cluster.go#L875), Check for changes by comparing current specs and newspecs. If no change found, return without any action.

**\- How to verify it**
Steps described in #25770 

**\- Description for the changelog**
- Avoids unnecessary service recreate for `docker service update` command. 

Fixes: #25770. Check for changes in old and new spec, Skip update
in case of no difference. This avoids recreate of service.

Signed-off-by: Kunal Kushwaha kushwaha_kunal_v7@lab.ntt.co.jp

### CODE REVIEW

File: daemon/cluster/cluster.go [hunk 1]
```diff
@@ -893,6 +893,17 @@ func (c *Cluster) UpdateService(serviceID string, version uint64, spec types.Ser
 		return err
 	}
 
+	// Check if anything is modified in spec for update #25770
+	currentService, err := getService(ctx, c.client, serviceID)
+	if err != nil {
+		return err
+	}
+
+	if serviceSpec.String() == currentService.String() {
```
Comment (AkihiroSuda) [inline daemon/cluster/cluster.go]: `serviceSpec` is `swarmapi.ServiceSpec`:
https://github.com/kunalkushwaha/docker/blob/ad6a1e09b2f7d2af01006e71c8f5542c71fffdcb/daemon/cluster/cluster.go#L834
https://github.com/kunalkushwaha/docker/blob/ad6a1e09b2f7d2af01006e71c8f5542c71fffdcb/daemon/cluster/convert/service.go#L96

However, `currentService` is `*swarmapi.Service`: https://github.com/kunalkushwaha/docker/blob/ad6a1e09b2f7d2af01006e71c8f5542c71fffdcb/daemon/cluster/helpers.go#L53

So I wonder if this comparison is robust.

Also, I fear the comparison of `String()` can fail intermittently if the structure contains map in future versions.

**Comment by cpuguy83** [comment]
This should be (and I think is) handled in the swarmkit side.

**Comment by kunalkushwaha** [comment]
@cpuguy83 I figured out the code in swarmkit.
But this [Update Function](https://github.com/docker/swarmkit/blob/master/manager/orchestrator/updater.go#L52)  is not working as per expected.

I tried to ``update.Cancel()`` if Task is not Dirty. But still the ``task.desiredstate=SHUTDOWN`` is assigned.

Below is log just after update command.  Will debug it further and update.

```
DEBU[0106] >>>> Specs Tasks matches                     
DEBU[0106] >>>> Specs Tasks Endpoint matches            
DEBU[0106] >>>>> service is not dirty...      
DEBU[0106] (*worker).Assign                              len(tasks)=1 module=node/agent
DEBU[0106] assigned                                      module=node/agent task.desiredstate=SHUTDOWN task.id=1gujwe3uvtyh5tzmrraybu9oe
ERRO[0106] container status unavailable                  error=context canceled module=node/agent/taskmanager task.id=1gujwe3uvtyh5tzmrraybu9oe
DEBU[0106] Assigning to node dndxr7zchlvr6kegfg3zcd9to   task.id=161ww84gu5pmr40fblyuz6jef
DEBU[0106] (*Agent).UpdateTaskStatus                     module=node/agent task.id=1gujwe3uvtyh5tzmrraybu9oe
DEBU[0106] task status reported                          module=node/agent
DEBU[0106] task status updated                           method=(*Dispatcher).processUpdates module=dispatcher state.transition=RUNNING->RUNNING task.id=1gujwe3uvtyh5tzmrraybu9oe
DEBU[0106] (*worker).Assign                              len(tasks)=2 module=node/agent
DEBU[0106] assigned                                      module=node/agent task.desiredstate=SHUTDOWN task.id=1gujwe3uvtyh5tzmrraybu9oe
DEBU[0106] assigned                                      module=node/agent task.desiredstate=READY task.id=161ww84gu5pmr40fblyuz6jef
DEBU[0106] state changed                                 module=node/agent state.desired=READY state.transition=ASSIGNED->ACCEPTED task.id=161ww84gu5pmr40fblyuz6jef
DEBU[0106] (*Agent).UpdateTaskStatus                     module=node/agent task.id=161ww84gu5pmr40fblyuz6jef
DEBU[0106] state changed                                 module=node/agent/taskmanager state.desired=READY state.transition=ASSIGNED->ACCEPTED task.id=161ww84gu5pmr40fblyuz6jef
DEBU[0106] task status reported                          module=node/agent
DEBU[0106] (*Agent).UpdateTaskStatus                     module=node/agent task.id=161ww84gu5pmr40fblyuz6jef
DEBU[0106] task status reported                          module=node/agent
DEBU[0106] state changed                                 module=node/agent/taskmanager state.desired=READY state.transition=ACCEPTED->PREPARING task.id=161ww84gu5pmr40fblyuz6jef
DEBU[0106] (*Agent).UpdateTaskStatus                     module=node/agent task.id=161ww84gu5pmr40fblyuz6jef
DEBU[0106] task status reported                          module=node/agent
DEBU[0106] task status updated                           method=(*Dispatcher).processUpdates module=dispatcher state.transition=ASSIGNED->PREPARING task.id=161ww84gu5pmr40fblyuz6jef

```

File: daemon/cluster/cluster.go (hunk 1 above)
Comment (kunalkushwaha) [inline daemon/cluster/cluster.go]: @AkihiroSuda Agree Strings in general not optimal for comparing. But ``serviceSpec.String()`` https://github.com/docker/swarmkit/blob/master/api/specs.pb.go#L2059 is swarmkit Function , so used.

I tried using reflect.DeepEqual() which was failing(coudn't verify why), so used this.

**Comment by tonistiigi** [comment]
@kunalkushwaha This case should be covered by https://github.com/docker/docker/blob/1dc87122ccb69741a6de2c933aed2d48b0b482a8/integration-cli/docker_api_swarm_test.go#L877 . If the rolling update happens I think it's because there is an actual change in spec(maybe from cli or conversion to grpc types).

**Comment by aluzzardi** [comment]
This should be handled in SwarmKit rather than here. This fix has has a race condition since we can't perform atomic read/compare/swap on the Engine side for service specifications.

Also - a possible side effect is that today we rely on `docker service update` (same spec) in order to resume a paused rollout and this fix would prevent that. @aaronlehmann ?

**Comment by aaronlehmann** [comment]
> This should be handled in SwarmKit rather than here. This fix has has a race condition since we can't perform atomic read/compare/swap on the Engine side for service specifications.

> Also - a possible side effect is that today we rely on docker service update (same spec) in order to resume a paused rollout and this fix would prevent that. @aaronlehmann ?

I agree with these comments.

SwarmKit shouldn't start a rolling update unless the spec has changed. If that isn't working as expected, I'd be happy to help debug and fix the issue.

**Comment by kunalkushwaha** [comment]
@tonistiigi I think ``isServiceDirty()`` https://github.com/docker/swarmkit/blob/master/manager/orchestrator/updater.go#L387 checks if service is dirty or not.
In this particular scenario, the current specs and newServiceSpecs are same and ``isServiceDirty`` return ``false``, Still the service is shutdown and recreated.

So I think bug is in swarmkit, will update more findings after debugging.

My steps to reproduce this bugs are exactly same as #25770

**Comment by tonistiigi** [comment]
@kunalkushwaha I saved out daemon logs for the case described in #25770 https://gist.github.com/tonistiigi/c0f8864648d3509a333778d50b7399d7 . As you can see a new mount is added on every update that changes the spec. So the problem is in docker cli.

**Comment by kunalkushwaha** [comment]
@tonistiigi  Yes I agree. Actually, I made a PR https://github.com/docker/docker/pull/25974, which fixes the docker cli issue of adding redundant target instead of updating it.

But I was able to reproduce the issue with that patch, so I tried to work on this issue too.

**Comment by tonistiigi** [comment]
@kunalkushwaha I can't reproduce with #25974 . Can you post daemon logs?

**Comment by kunalkushwaha** [comment]
@tonistiigi I am also not able to reproduce this issue now with #25974 . I think I did something wrong while applying that patch :( . Sorry for the whole noise.

**Comment by tonistiigi** [comment]
I'll close this then. Let's continue the discussion in the issues and #25974 .

**Pull request closed without merging**